synapse/lib/modelrev.py

```diff
@@ -1205,3 +1206,745 @@ async def _propArrayToForm(self, layers, propfull, formname):
+
+        await self.delNode(buid)
+
+    async def storeNode(self, buid):
+        assert self.nodes.has(buid)
+        node = self.getNode(buid)
+
+        formname = node.get('formname')
+        formvalu = node.get('formvalu')
+
+        sources = set()
+        # Resolve sources
+        n2edges = {}
+        for layriden, edges in node['n2edges'].items():
+            n2edges.setdefault(layriden, [])
+
+            for verb, n2iden in edges:
+                n2buid = s_common.uhex(n2iden)
+                assert self.nodes.has(n2buid)
+                n2node = self.getNode(n2buid)
+
+                n2edges[layriden].append((verb, n2iden, n2node['formname']))
+
+                if verb == 'seen':
+                    formvalu = n2node.get('formvalu')
+                    assert formvalu is not None
+                    sources.add(formvalu)
+
+        # Make some changes before serializing
+        item = s_msgpack.deepcopy(node)
+        item.pop('verdict', None)
+        item['iden'] = s_common.ehex(buid)
+        item['sources'] = list(sources)
+        item['n2edges'] = n2edges
+
+        roprops = self.getRoProps(formname)
+        for layriden, sode in node['sodes'].items():
+            props = sode.get('props')
+            if props is None: # pragma: no cover
+                continue
+
+            props = {name: valu for name, valu in list(props.items()) if name not in roprops}
+            if props:
+                item['sodes'][layriden]['props'] = props
+            else:
+                item['sodes'][layriden].pop('props')
+
+        await self.core.coreQueuePuts('model_0_2_31:nodes', (item,))
+
+    async def getSodeByPropValuNoNorm(self, layer, formname, propname, valu, cmpr='='):
+        prop = self.core.model.reqProp(f'{formname}:{propname}')
+
+        stortype = prop.type.stortype
+
+        # Normally we'd call proptype.getStorCmprs() here to get the cmprvals
+        # but getStorCmprs() calls norm() which we're  trying to avoid so build
+        # cmprvals manually here.
+
+        if prop.type.isarray:
+            stortype &= (~s_layer.STOR_FLAG_ARRAY)
+            liftfunc = layer.liftByPropArray
+        else:
+            liftfunc = layer.liftByPropValu
+
+        cmprvals = ((cmpr, valu, stortype),)
+
+        async for _, buid, sode in liftfunc(formname, propname, cmprvals):
+            yield buid, sode
+
+    async def delNode(self, buid):
+        assert self.nodes.has(buid)
+        node = self.getNode(buid)
+
+        formname = node.get('formname')
+        formvalu = node.get('formvalu')
+
+        # Edits
+        for layriden, sode in node['sodes'].items():
+            props = sode.get('props', {}).copy()
+            for propname, propvalu in props.items():
+                propvalu, stortype = propvalu
+                await self.editPropDel(layriden, buid, formname, propname, propvalu, stortype)
+
+            tags = sode.get('tags', {}).copy()
```

Comment:
Do we actually need to iterate over copies of these objects here? Doesn't look like they get modified

Reply:
Props get modified but not the other info. I'll update.